osutil/io.go

```diff
@@ -62,53 +98,120 @@ func AtomicWriteFileChown(filename string, data []byte, perm os.FileMode, flags
+			return err
 		}
+	}
+
+	var dir *os.File
+	if !snapdUnsafeIO {
+		// XXX: if go switches to use aio_fsync, we need to open the dir for writing
+		d, err := os.Open(filepath.Dir(aw.target))
 		if err != nil {
-			os.Remove(tmp)
+			return err
 		}
-	}()
-
-	// according to the docs, Write returns a non-nil error when n !=
-	// len(b), so don't worry about short writes.
-	if _, err := fd.Write(data); err != nil {
-		return err
-	}
+		dir = d
+		defer dir.Close()
 
-	if uid > -1 && gid > -1 {
-		if err := fd.Chown(uid, gid); err != nil {
+		if err := aw.Sync(); err != nil {
 			return err
 		}
-	} else if uid > -1 || gid > -1 {
-		return errors.New("internal error: AtomicWriteFileChown needs none or both of uid and gid set")
 	}
 
+	if err := os.Rename(aw.tmpname, aw.target); err != nil {
+		return err
+	}
+	aw.renamed = true // it is now too late to Cancel()
+
 	if !snapdUnsafeIO {
-		if err := fd.Sync(); err != nil {
+		if err := dir.Sync(); err != nil {
 			return err
 		}
 	}
 
-	if err := os.Rename(tmp, filename); err != nil {
+	// given we called Sync before, Close _shouldn't_ be able to
+	// fail. Still, stuff happens.
+	return aw.Close()
+}
+
+// The AtomicWrite* family of functions work like ioutil.WriteFile(), but the
+// file created is an AtomicWriter, which is Finalized before returning.
+//
+// AtomicWriteChown and AtomicWriteFileChown take an uid and a gid that can be
+// used to specify the ownership of the created file. They must be both
+// non-negative (in which case chown is called), or both negative (in which
+// case it isn't).
+//
+// AtomicWriteFile and AtomicWriteFileChown take the content to be written as a
+// []byte, and so work exactly like io.WriteFile(); AtomicWrite and
+// AtomicWriteChown take an io.Reader which is copied into the file instead,
+// and so are more amenable to streaming.
+func AtomicWrite(filename string, reader io.Reader, perm os.FileMode, flags AtomicWriteFlags) (err error) {
+	return AtomicWriteChown(filename, reader, perm, flags, -1, -1)
+}
+
+func AtomicWriteFile(filename string, data []byte, perm os.FileMode, flags AtomicWriteFlags) (err error) {
+	return AtomicWriteChown(filename, bytes.NewReader(data), perm, flags, -1, -1)
+}
+
+func AtomicWriteFileChown(filename string, data []byte, perm os.FileMode, flags AtomicWriteFlags, uid, gid int) (err error) {
+	return AtomicWriteChown(filename, bytes.NewReader(data), perm, flags, uid, gid)
+}
+
+func AtomicWriteChown(filename string, reader io.Reader, perm os.FileMode, flags AtomicWriteFlags, uid, gid int) (err error) {
+	aw, err := NewAtomicFile(filename, perm, flags, uid, gid)
+	if err != nil {
 		return err
 	}
 
-	if !snapdUnsafeIO {
-		return dir.Sync()
+	defer func() {
+		if err != nil {
+			// XXX there's a small window where Finalize can fail in the last two steps
+			// (syncing the containing directory, or closing the file), and this quietly
+			// ignores that -- not that there'd be much we could do!
```

Comment:
this is confusing, it sound like we wouldn't return an error but we do in that case out of the function?

Reply:
it's a bad enough situation I'm tempted to panic, here. I've been thinking about ways to make it better and I think I can move the Close to before the Rename, meaning the only thing failing is the sync on the directory, so it should be fine then.

Reply:
FWIW: yes we're returning an error, but if something threw an error and then the Cancel fails, the whole thing is inconsistent. You don't know what you have.

Reply:
(and, to note, the old code had the same problem but it wasn't apparent)